editwatcher.py

```diff
@@ -0,0 +1,160 @@
+# noinspection PyPackageRequirements
+import websocket
+# noinspection PyPackageRequirements
+from bs4 import BeautifulSoup
+from urllib.parse import urlparse
+import chatcommunicate
+import metasmoke
+from globalvars import GlobalVars
+import datahandling
+from helpers import log, add_to_global_bodyfetcher_queue_in_new_thread
+from parsing import fetch_post_id_and_site_from_url, to_protocol_relative
+from tasks import Tasks
+
+PICKLE_FILENAME = "editActions.p"
+DEFAULT_TIMEOUT = 10 * 60  # 10 minutes
+
+
+# noinspection PyClassHasNoInit,PyBroadException,PyMethodParameters
+class EditWatcher:
+    def __init__(self):
+        if GlobalVars.no_edit_watcher:
+            self.socket = None
+            return
+        # posts is a dict with the WebSocket action as keys {site_id}-question-{question_id} as keys
+        # with each value being: (site_id, hostname, question_id, max_time)
+        self.posts = {}
+        self.posts_lock = threading.Lock()
+        self.save_handle = None
+        self.save_handle_lock = threading.Lock()
+
+        try:
+            self.socket = websocket.create_connection("wss://qa.sockets.stackexchange.com/")
+        except websocket.WebSocketException:
+            self.socket = None
+            log('error', 'EditWatcher failed to create a websocket connection')
+
+        if datahandling.has_pickle(PICKLE_FILENAME):
+            pickle_data = datahandling.load_pickle(PICKLE_FILENAME)
+            now = time.time()
+            new_posts = {action: value for action, value in pickle_data if value[-1] > now}
+            with self.posts_lock:
+                self.posts = new_posts
+            for action in new_posts.keys():
+                Tasks.do(self._subscribe, action)
+            self._schedule_save()
+
+        threading.Thread(name="edit watcher", target=self._start, daemon=True).start()
+
+    def _start(self):
+        while True:
+            msg = self.socket.recv()
+
+            if msg:
+                msg = json.loads(msg)
+                action = msg["action"]
+
+                if action == "hb":
+                    self.socket.send("hb")
+                else:
+                    data = json.loads(msg["data"])
+                    now = time.time()
+                    with self.posts_lock:
+                        site_id, hostname, question_id, max_time = self.posts.get(action, (None, None, None, now))
+                        if site_id and max_time <= now:
+                            del self.posts[action]
+                            Tasks.do(self._unsubscribe, action)
+                    if max_time > now and data["a"] == "post-edit":
+                        add_to_global_bodyfetcher_queue_in_new_thread(hostname, question_id, False)
+
+    def subscribe(self, post_url=None, hostname=None, site_id=None, question_id=None,
+                  pickle=True, timeout=DEFAULT_TIMEOUT, max_time=None, from_time=None):
+        if GlobalVars.no_edit_watcher:
+            return
```

Comment:
Maybe add INFO logging here to make it discoverable when the code goes no-op?

Reply:
This `if` should only ever trigger based on the SD owner explicitly providing the `no_edit_watcher` option on the command line when invoking nocrash.py. Logging that here, and in deletionwatcher.py, which does the same thing, would get really noisy for the cases where those are disabled. Those cases are fairly likely to be when someone is doing testing and actually looking at the console logging, so would have an impact on debugging. I'd be OK with an INFO log where we decode the command line and set those values in `GlobalVars`. That wouldn't be all *that* noisy, but might be annoying, given that `EditWatcher` and `DeletionWatcher` are only ever disabled when the user has explicitly specified it in the command line. While those command line options could be used for other purposes (e.g. they are used for the metasmoke instance), they were primarily added for testing.

Reply:
Yup, no objection. I haven't tested this code but I have been in similar scenarios with other code where a simple log statement would have saved me some digging to figure out that I was disabling a code flow somewhere deep inside.